### PR TITLE
source-mysql: Use DB `time_zone` for DATETIME parsing

### PR DESCRIPTION
**Pull request by willdonnelly**
**Description:**

The DATETIME column type is just a raw 'YYYY-MM-DD HH:MM:SS' value with no explicit frame of reference. In general there's an implicit reference of "the user's local time zone", but we don't necessarily know what that is. And we'd really like to emit captured values as RFC3339 timestamps, so we need to interpret the values *somewhere*.

After this PR we assume that if the database's `time_zone` system variable is set to a IANA zone name or a numeric time offset then that is the location/offset in which a DATETIME should be parsed.

If the `time_zone` variable is set to its default value of `'SYSTEM'` we refuse to translate DATETIMEs, because for some reason the MySQL `system_time_zone` variable only contains the current time offset [1] and not a valid zone name which could be used to interpret historical/future dates.

[1] To be concrete: if the system locale is set to 'America/Los_Angeles' then MySQL's `system_time_zone` will report 'PST' or 'PDT' at different times of the year, rather than something like 'PST8PDT' or 'America/Los_Angeles' which could actually be used to interpret historical data. In addition the strings 'PST' and 'PDT' are not valid IANA time zone names, so we just avoid all that complexity by requiring a valid `time_zone` setting.

**Documentation links affected:**

We should update https://docs.estuary.dev/reference/Connectors/capture-connectors/MySQL/ to discuss our requirement that the `time_zone` system variable be set to a valid time zone name or numeric offset if the user wants to capture `DATETIME` columns. Then https://go.estuary.dev/80J6rX should be updated to link to that particular bit of the docs.

- - -
This change is [Reviewable](https://reviewable.io/reviews/estuary/connectors/440)

### CODE REVIEW

File: source-mysql/main.go [hunk 1]
```diff
@@ -245,22 +258,57 @@ func (db *mysqlDatabase) connect(ctx context.Context) error {
...
+		}
+		minutes, err := strconv.Atoi(parts[1])
+		if err != nil {
+			return nil, fmt.Errorf("error parsing %q: %w", tzName, err)
+		}
+		return time.FixedZone("UTC"+tzName, sign*(hours*60*60+minutes*60)), nil
+	}
+
+	// Otherwise let's hope the time zone setting is a valid IANA zone name.
+	return time.LoadLocation(tzName)
```
Comment (mdibaiee) [inline source-mysql/main.go:293]: I kinda want to have an easter egg that says "Thank You" if the timezone is set to UTC 😂

**Review by mdibaiee — APPROVED** [review]

File: source-mysql/main.go [hunk 2]
```diff
@@ -223,7 +227,16 @@ func (db *mysqlDatabase) connect(ctx context.Context) error {
...
+	if loc, err := queryTimeZone(conn); err == nil {
+		db.datetimeLocation = loc
+		logrus.WithField("loc", loc.String()).Debug("using datetime location")
+	} else {
+		logrus.WithField("err", err).Warn("unable to determine database timezone")
+		logrus.Warn("capturing DATETIME values will not be permitted")
+	}
+
+	// Set our desired timezone to UTC. This is required for backfills of TIMESTAMP columns to behave consistently.
 	if _, err := db.conn.Execute("SET time_zone = '+00:00';"); err != nil {
```
Comment (jgraettinger) [inline source-mysql/main.go:240]: Making sure I'm following: we're fetching a persisted `time_zone` on startup, and then overwriting it _for our session only_ as part of making backfills work for TIMESTAMP? This is sound because DATETIME query results are not affected by setting `time_zone`, while TIMESTAMP will consistently be UTC, and no other sessions will observe our write to `time_zone`?

File: source-mysql/main.go [hunk 3]
```diff
@@ -245,22 +258,57 @@ func (db *mysqlDatabase) connect(ctx context.Context) error {
...
+	if err != nil {
+		return nil, fmt.Errorf("error querying 'time_zone' system variable: %w", err)
+	}
+	logrus.WithField("time_zone", tzName).Debug("queried time_zone system variable")
+	if tzName == "SYSTEM" {
+		return nil, errDatabaseTimezoneUnknown
+	}
+
+	// If it looks like a numeric offset then construct a fixed-offset time zone from that.
+	if timeZoneOffsetRegex.MatchString(tzName) {
```
Comment (jgraettinger) [inline source-mysql/main.go:274]: Should we just ... not have this behavior? It's extra code to maintain that could have subtle bugs, it's more documentation to write and to communicate to users, and it doesn't seem unreasonable to simply require an IANA time zone name.

If we do keep it, it looks like it needs test coverage (only IANA case is covered currently)

**Review by jgraettinger — COMMENTED** [review]
Looking great, just a question & comment. Heads up there's a test failure in CI as well

File: source-mysql/main.go (hunk 2 above)
Comment (willdonnelly) [reply source-mysql/main.go:240]: That sounds exactly right. The read is of the global setting of `time_zone` which should apply to all sessions by default, and then we modify it for just our session so that the DB will do the right thing and give us `TIMESTAMP` values in UTC.

File: source-mysql/main.go (hunk 3 above)
Comment (willdonnelly) [reply source-mysql/main.go:274]: I included this logic because it seemed reasonable to support any legal setting of the `time_zone` variable that wasn't `'SYSTEM'`.

I'm not super attached to keeping it, but it feels worthwhile to me to support as many customer database configurations as we reasonably can, and parsing the numeric offset isn't super complicated (and in fact now that I'm well rested I've realized we can get Go to do the parsing with something like `t, _ := time.Parse("-07:00", tzName); t.Location()` so that avoids hand-written parsing code here).

So I'm inclined to keep this support but replace the hand-written parser with Go's `time.Parse` and then add a new test case which exercises DATETIME support in more time zones than just `America/Chicago`.

File: source-mysql/main.go (hunk 2 above)
Comment (willdonnelly) [reply source-mysql/main.go:241]: I've tweaked the comment to hopefully clarify what's going on for future readers.

**Comment by willdonnelly** [comment]
Currently looking into the CI failure. It looks like tests exercising `TIMESTAMP` round-tripping are failing, for what is in hindsight a fairly obvious reason: the tests set the **global** time zone but don't modify the **session** time zone of the connection used for inserting test data, and so that just inherits whatever the server's global time zone was when we launched the current test suite run.

And I never noticed that locally because after running a particular sub-test once the global setting is correct for that specific test if you run it again against the same database. So I think the fix is just to set the session timestamp (for the test-data-inserting connection, which is distinct from the capture backfill connection that we set to UTC so the capture behaves properly. So many distinct variables...)

**Comment by willdonnelly** [comment]
Looks like everything is passing CI now, and with the new `TestDatetimes` test I'm fairly confident we've actually got this correct. Let me know if there are any time zone edge cases that ought to be added there, currently I've got:

- Named time zone which observes DST (with dates in both the summer and the winter)
- Named time zone which does not observe DST
- Explicit offset time zone

**Review by jgraettinger — APPROVED** [review]
LGTM!
